x/wasm/keeper/relay.go

```diff
@@ -130,13 +133,34 @@ func (k Keeper) OnRecvPacket(
 	res, gasUsed, execErr := k.wasmVM.IBCPacketReceive(codeInfo.CodeHash, env, msg, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas, costJSONDeserialization)
 	k.consumeRuntimeGas(ctx, gasUsed)
 	if execErr != nil {
-		return nil, errorsmod.Wrap(types.ErrExecuteFailed, execErr.Error())
+		panic(execErr) // let contract fully abort IBC receive in certain case
 	}
-	if res.Err != "" { // handle error case as before https://github.com/CosmWasm/wasmvm/commit/c300106fe5c9426a495f8e10821e00a9330c56c6
-		return nil, errorsmod.Wrap(types.ErrExecuteFailed, res.Err)
+	if res.Err != "" {
+		// return error ACK with non-redacted contract message
+		return channeltypes.Acknowledgement{
+			Response: &channeltypes.Acknowledgement_Error{Error: res.Err},
+		}, nil
 	}
 	// note submessage reply results can overwrite the `Acknowledgement` data
-	return k.handleContractResponse(ctx, contractAddr, contractInfo.IBCPortID, res.Ok.Messages, res.Ok.Attributes, res.Ok.Acknowledgement, res.Ok.Events)
+	data, err := k.handleContractResponse(ctx, contractAddr, contractInfo.IBCPortID, res.Ok.Messages, res.Ok.Attributes, res.Ok.Acknowledgement, res.Ok.Events)
+	if err != nil {
+		// submessage errors result in error ACK
+		return nil, err
+	}
+	// success ACK
+	return ContractConfirmStateAck(data), nil
```

Comment:
I think this code or comment is misleading because it can be either a success or and error or a non-binary acknowledgement type. We should not try to give it any meaning.

Reply:
In this line Success can be defined as case 1 + 2 of https://github.com/CosmWasm/wasmd/issues/697#issuecomment-992573516 . The contract was able to fully handle the packet. State will be committed.